### PR TITLE
feat-new-db-query

### PR DESCRIPTION
**Pull request by lohanidamodar**
New class for updated database query parameters

### CODE REVIEW

File: templates/dart/lib/services/service.dart.twig [hunk 1]
```diff
@@ -1,11 +1,16 @@
 part of {{ language.params.packageName }};
 
 {% macro parameter(parameter) %}
-{% if parameter.required %}required {{ parameter.type | typeName }}{% else %}{{ parameter.type | typeName }}?{% endif %} {{ parameter.name | caseCamel }}{% endmacro %}
+{% if parameter.name == 'queries' %}Query? queries{% else %}
```
Comment (TorstenDittmann) [inline templates/dart/lib/services/service.dart.twig:4]: How safe is depending on the name itself for us? 🤔

**Review by TorstenDittmann — APPROVED** [review]
Looks good to me 👍🏻

File: templates/dart/lib/services/service.dart.twig (hunk 1 above)
Comment (lohanidamodar) [reply templates/dart/lib/services/service.dart.twig:4]: we did it before for orderType, but we removed that enum

**Review by eldadfux — CHANGES_REQUESTED** [review]
I don't feel comfortable at all with relying on the param name for using the new Query filter. We can and need to change the swagger format so we add and x-something attribute to indicate this param has special capabilities.

**Comment by lohanidamodar** [comment]
@eldadfux This is how the listDocuments method looks
```dart
Future<Response> listDocuments({required String collectionId, Query? queries, int? limit, int? offset, List? orderAttributes, List? orderTypes}) {
        final String path = '/database/collections/{collectionId}/documents'.replaceAll(RegExp('{collectionId}'), collectionId);

        final Map<String, dynamic> params = {
            'queries': queries?.queries,
            'limit': limit,
            'offset': offset,
            'orderAttributes': orderAttributes,
            'orderTypes': orderTypes,
        };

        final Map<String, String> headers = {
            'content-type': 'application/json',
        };

        return client.call(HttpMethod.get, path: path, params: params, headers: headers);
    }
```

File: templates/dart/lib/query.dart.twig [hunk 2]
```diff
@@ -0,0 +1,50 @@
+part of {{ language.params.packageName }};
+
+class Query {
```
Comment (lohanidamodar) [inline templates/dart/lib/query.dart.twig:3]: @eldadfux this is how the query class looks

**Comment by eldadfux** [comment]
> @eldadfux This is how the listDocuments method looks
> 
> ```dart
> Future<Response> listDocuments({required String collectionId, Query? queries, int? limit, int? offset, List? orderAttributes, List? orderTypes}) {
>         final String path = '/database/collections/{collectionId}/documents'.replaceAll(RegExp('{collectionId}'), collectionId);
> 
>         final Map<String, dynamic> params = {
>             'queries': queries?.queries,
>             'limit': limit,
>             'offset': offset,
>             'orderAttributes': orderAttributes,
>             'orderTypes': orderTypes,
>         };
> 
>         final Map<String, String> headers = {
>             'content-type': 'application/json',
>         };
> 
>         return client.call(HttpMethod.get, path: path, params: params, headers: headers);
>     }
> ```

Would be great if you could also share an example of how it will be called with multiple queries.

**Comment by lohanidamodar** [comment]
> > @eldadfux This is how the listDocuments method looks
> > ```dart
> > Future<Response> listDocuments({required String collectionId, Query? queries, int? limit, int? offset, List? orderAttributes, List? orderTypes}) {
> >         final String path = '/database/collections/{collectionId}/documents'.replaceAll(RegExp('{collectionId}'), collectionId);
> > 
> >         final Map<String, dynamic> params = {
> >             'queries': queries?.queries,
> >             'limit': limit,
> >             'offset': offset,
> >             'orderAttributes': orderAttributes,
> >             'orderTypes': orderTypes,
> >         };
> > 
> >         final Map<String, String> headers = {
> >             'content-type': 'application/json',
> >         };
> > 
> >         return client.call(HttpMethod.get, path: path, params: params, headers: headers);
> >     }
> > ```
> 
> Would be great if you could also share an example of how it will be called with multiple queries.

```dart
db.listDocuments(collectionId: 'test', queries: Query().equal('director','Charles').contains('cast', ['Arnold', 'Vin']).lesser('release_year', 2005), limit: 5);
```

**Comment by lohanidamodar** [comment]
@eldadfux I just remembered, we could add a build method in the query class that will return the queries as list of strings so above example can be like
```
db.listDocuments(
    collectionId: 'test',
    queries: Query()
        .equal('director','Charles')
        .contains('cast', ['Arnold', 'Vin'])
        .lesser('release_year', 2005)
        .build(),
    limit: 5,
);
```
Doing this, I could change the listDocuments to 

```dart
Future<Response> listDocuments({required String collectionId, List<String> queries, int? limit, int? offset, List? orderAttributes, List? orderTypes}) {
        final String path = '/database/collections/{collectionId}/documents'.replaceAll(RegExp('{collectionId}'), collectionId);

        final Map<String, dynamic> params = {
            'queries': queries,
            'limit': limit,
            'offset': offset,
            'orderAttributes': orderAttributes,
            'orderTypes': orderTypes,
        };

        final Map<String, String> headers = {
            'content-type': 'application/json',
        };

        return client.call(HttpMethod.get, path: path, params: params, headers: headers);
    }
```

**Comment by eldadfux** [comment]
@lohanidamodar @TorstenDittmann @kodumbeats what approach would you like better:

1. shorter syntax, easier to prepare internally and convert to an HTTP call: 
```dart
db.listDocuments(
    collectionId: 'test',
    queries: Query()
        .equal('director','Charles')
        .contains('cast', ['Arnold', 'Vin'])
        .lesser('release_year', 2005),
    limit: 5,
);
```

2. Longer, but more verbose, better relationship between `queries` and the `Query` class. More obvious this is an array.

```dart
db.listDocuments(
    collectionId: 'test',
    queries: [
        Query().equal('director','Charles'),
        Query().contains('cast', ['Arnold', 'Vin']),
        Query().lesser('release_year', 2005),
    ]
    limit: 5,
);
```

**Comment by lohanidamodar** [comment]
With the new query class calling listDocuments looks like
```dart
db.listDocuments(
    collectionId: 'test',
    queries: [
        Query.equal('director','Charles'),
        Query.contains('cast', ['Arnold', 'Vin']),
        Query.lesser('release_year', 2005),
    ]
    limit: 5,
);
```

File: templates/web/src/query.ts.twig [hunk 3]
```diff
@@ -0,0 +1,37 @@
+class Query {
+  static equal(attribute: string, value: any): string {
```
Comment (TorstenDittmann) [inline templates/web/src/query.ts.twig:2]: I would replace `any` with the actual types we allow 🤔 

Lets add a type like this 
```ts
type QueryTypes = string | string[] | number | number[] | boolean | boolean[];
```

and we can use it like this
```ts
static equal(attribute: string, value: QueryTypes) {

}
```

File: templates/web/src/query.ts.twig [hunk 4]
```diff
@@ -0,0 +1,37 @@
...
+
+  static greaterEqual(attribute: string, value: any): string {
+    return Query.addQuery(attribute, 'greaterEqual', value);
+  }
+
+  static contains(attribute: string, value: any): string {
+    return Query.addQuery(attribute, 'contains', value);
+  }
+
+  static addQuery(attribute: string, oper: string, value: any): string {
```
Comment (TorstenDittmann) [inline templates/web/src/query.ts.twig:30]:
```suggestion
  private static addQuery(attribute: string, oper: string, value: any): string {
```

File: templates/web/src/query.ts.twig [hunk 5]
```diff
@@ -0,0 +1,37 @@
...
+    if (typeof value === 'string') {
```
Comment (TorstenDittmann) [inline templates/web/src/query.ts.twig:31]: How de we handle arrays here?

Comment (lohanidamodar) [reply templates/web/src/query.ts.twig:31]: should work automatically.

Comment (TorstenDittmann) [reply templates/web/src/query.ts.twig:34]: I don't think so.

```ts
Query.equal("attribute", ["test", "with,comma"])
```
will result in 
```ts
'attribute.equal(test,with,comma)'
```

File: templates/android/library/src/main/java/io/appwrite/Query.kt.twig [hunk 6]
```diff
@@ -0,0 +1,41 @@
...
+    fun contains(attribute: String, value: List<Any>): String {
+      return Query.addQuery(attribute, "contains", value);
+    }
+
+    fun addQuery(attribute: String, oper: String, value: Any): String {
+      if (value is String) {
+        return "${attribute}.${oper}(\"${value}\")";
+      } else {
+        return "${attribute}.${oper}(${value})";
+      }
```
Comment (abnegate) [inline templates/android/library/src/main/java/io/appwrite/Query.kt.twig]:
```suggestion
      return when (value) {
         is String -> "${attribute}.${oper}(\"${value}\")"
         else -> "${attribute}.${oper}(${value})"
       }
```

File: templates/android/library/src/main/java/io/appwrite/Query.kt.twig [hunk 7]
```diff
@@ -0,0 +1,41 @@
+package {{ sdk.namespace | caseDot }}
+
+class Query {
+  companion object {
+    fun equal(attribute: String, value: Any): String {
+      return Query.addQuery(attribute, "equal", value);
+    }
```
Comment (abnegate) [inline templates/android/library/src/main/java/io/appwrite/Query.kt.twig:7]: These functions could be expression bodied:

```suggestion
    fun equal(attribute: String, value: Any) -> addQuery(attribute, "equal", value)
```

**Review by TorstenDittmann — APPROVED** [review]
Web and Deno lgtm 👍🏻

**Pull request merged**